File: plugins/logistics/backend/services/planning.py
from __future__ import annotations

from datetime import date

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from plugins.logistics.backend.common import (
    LogisticsActionContext,
    audit_logistics_action,
    emit_logistics_event,
)
from plugins.logistics.backend.models import (
    LogisticsMovement,
    LogisticsMovementItem,
    LogisticsOrder,
    LogisticsOrderItem,
    LogisticsPlanPreload,
    LogisticsPlanPreloadItem,
    LogisticsWarehouse,
)
from plugins.logistics.backend.schemas import (
    MovementRead,
    OrderItemRead,
    PlanningGeneratePreloadRequest,
    PlanningPendingOrderItemRead,
    PlanningPendingOrderRead,
    PlanningPlanOrderRequest,
    PlanningPlanOrderResult,
    PlanningPreloadActionResult,
    PlanningPreloadItemRead,
    PlanningPreloadRead,
    PlanningStockSummaryItem,
)
from plugins.productos.backend.models import Product
from plugins.stock.backend.models import StockBalance

OPEN_ORDER_STATUSES = ("PENDIENTE", "PLANIFICADO")
ACTIVE_PRELOAD_STATUSES = ("PENDIENTE", "ACEPTADA")


def _coverage_status(*, available: float, required: float) -> str:
    if required <= 0:
        return "green"
    if available >= required:
        return "green"
    if available > 0:
        return "yellow"
    return "red"


def _product_name_map(db: Session, *, tenant_id: str, product_ids: set[str]) -> dict[str, str]:
    if not product_ids:
        return {}
    rows = db.execute(
        select(Product.id, Product.name).where(
            Product.tenant_id == tenant_id,
            Product.id.in_(product_ids),
        )
    ).all()
    return {row.id: row.name for row in rows}


def _stock_actual_by_product(
    db: Session,
    *,
    tenant_id: str,
    warehouse_id: str,
    product_ids: set[str],
) -> dict[str, float]:
    if not product_ids:
        return {}
    rows = db.execute(
        select(StockBalance.product_id, func.coalesce(func.sum(StockBalance.quantity), 0))
        .where(
            StockBalance.tenant_id == tenant_id,
            StockBalance.warehouse_id == warehouse_id,
            StockBalance.product_id.in_(product_ids),
        )
        .group_by(StockBalance.product_id)
    ).all()
    return {row[0]: float(row[1] or 0) for row in rows}


def _stock_comprometido_by_product(
    db: Session,
    *,
    tenant_id: str,
    warehouse_id: str,
    product_ids: set[str],
    exclude_order_id: str | None = None,
) -> dict[str, float]:
    if not product_ids:
        return {}
    stmt = (
        select(
            LogisticsOrderItem.product_id,
            func.coalesce(func.sum(LogisticsOrderItem.quantity_planned), 0),
        )
        .join(LogisticsOrder, LogisticsOrder.id == LogisticsOrderItem.order_id)
        .where(
            LogisticsOrder.tenant_id == tenant_id,
            LogisticsOrder.warehouse_id == warehouse_id,
            LogisticsOrder.status.in_(OPEN_ORDER_STATUSES),
            LogisticsOrderItem.product_id.is_not(None),
            LogisticsOrderItem.product_id.in_(product_ids),
        )
        .group_by(LogisticsOrderItem.product_id)
    )
    if exclude_order_id is not None:
        stmt = stmt.where(LogisticsOrder.id != exclude_order_id)
    rows = db.execute(stmt).all()
    return {row[0]: float(row[1] or 0) for row in rows}


def _stock_planificado_by_product(
    db: Session,
    *,
    tenant_id: str,
    warehouse_id: str,
    product_ids: set[str],
) -> dict[str, float]:
    if not product_ids:
        return {}
    rows = db.execute(
        select(
            LogisticsPlanPreloadItem.product_id,
            func.coalesce(func.sum(LogisticsPlanPreloadItem.quantity_planned), 0),
        )
        .join(LogisticsPlanPreload, LogisticsPlanPreload.id == LogisticsPlanPreloadItem.preload_id)
        .where(
            LogisticsPlanPreload.tenant_id == tenant_id,
            LogisticsPlanPreload.warehouse_id == warehouse_id,
            LogisticsPlanPreload.status == "PENDIENTE",
            LogisticsPlanPreloadItem.product_id.in_(product_ids),
        )
        .group_by(LogisticsPlanPreloadItem.product_id)
    ).all()
    return {row[0]: float(row[1] or 0) for row in rows}


def list_stock_summary(
    db: Session,
    *,
    tenant_id: str,
    warehouse_id: str,
    product_ids: set[str],
) -> list[PlanningStockSummaryItem]:
    names = _product_name_map(db, tenant_id=tenant_id, product_ids=product_ids)
    actual = _stock_actual_by_product(
        db, tenant_id=tenant_id, warehouse_id=warehouse_id, product_ids=product_ids
    )
    comprometido = _stock_comprometido_by_product(
        db, tenant_id=tenant_id, warehouse_id=warehouse_id, product_ids=product_ids
    )
    planificado = _stock_planificado_by_product(
        db, tenant_id=tenant_id, warehouse_id=warehouse_id, product_ids=product_ids
    )
    items: list[PlanningStockSummaryItem] = []
    for product_id in sorted(product_ids):
        stock_actual = actual.get(product_id, 0)
        stock_comprometido = comprometido.get(product_id, 0)
        stock_planificado = planificado.get(product_id, 0)
        stock_disponible = stock_actual - stock_comprometido - stock_planificado
        items.append(
            PlanningStockSummaryItem(
                product_id=product_id,
                product_name=names.get(product_id, product_id),
                warehouse_id=warehouse_id,
                stock_actual=stock_actual,
                stock_comprometido=stock_comprometido,
                stock_planificado=stock_planificado,
                stock_disponible=stock_disponible,
                coverage_status=_coverage_status(available=stock_disponible, required=1),
            )
        )
    return items


def list_pending_orders(
    db: Session,
    *,
    tenant_id: str,
    allowed_warehouse_ids: tuple[str, ...] | None,
    warehouse_id: str | None = None,
) -> list[PlanningPendingOrderRead]:
    stmt = select(LogisticsOrder).where(
        LogisticsOrder.tenant_id == tenant_id,
        LogisticsOrder.status.in_(OPEN_ORDER_STATUSES),
    )
    if warehouse_id is not None:
        stmt = stmt.where(LogisticsOrder.warehouse_id == warehouse_id)
    if allowed_warehouse_ids is not None:
        stmt = stmt.where(LogisticsOrder.warehouse_id.in_(allowed_warehouse_ids))
    orders = list(db.scalars(stmt.order_by(LogisticsOrder.created_at.desc())).all())
    result: list[PlanningPendingOrderRead] = []
    for order in orders:
        items = list(
            db.scalars(
                select(LogisticsOrderItem)
                .where(LogisticsOrderItem.order_id == order.id)
                .order_by(LogisticsOrderItem.created_at.asc())
            ).all()
        )
        product_ids = {item.product_id for item in items if item.product_id is not None}
        available_map = (
            {
                item.product_id: item.stock_disponible
                for item in list_stock_summary(
                    db,
                    tenant_id=tenant_id,
                    warehouse_id=order.warehouse_id or "",
                    product_ids=product_ids,
                )
            }
            if order.warehouse_id is not None
            else {}
        )
        item_reads: list[PlanningPendingOrderItemRead] = []
        order_status = "green"
        for item in items:
            pending = max(float(item.quantity_requested) - float(item.quantity_planned), 0)
            available = available_map.get(item.product_id, 0) if item.product_id is not None else 0
            coverage = _coverage_status(available=available, required=pending)
            if coverage == "red":
                order_status = "red"
            elif coverage == "yellow" and order_status != "red":
                order_status = "yellow"
            item_reads.append(
                PlanningPendingOrderItemRead(
                    order_item_id=item.id,
                    product_id=item.product_id,
                    product_name=item.product_name,
                    quantity_requested=float(item.quantity_requested),
                    quantity_planned=float(item.quantity_planned),
                    quantity_pending=pending,
                    stock_disponible=available,
                    coverage_status=coverage,
                )
            )
        result.append(
            PlanningPendingOrderRead(
                order_id=order.id,
                customer_id=order.customer_id,
                customer_name=order.customer_name,
                warehouse_id=order.warehouse_id,
                status=order.status,
                coverage_status=order_status,
                items=item_reads,
            )
        )
    return result


def _get_preload_items(db: Session, *, preload_id: str) -> list[LogisticsPlanPreloadItem]:
    return list(
        db.scalars(
            select(LogisticsPlanPreloadItem)
            .where(LogisticsPlanPreloadItem.preload_id == preload_id)
            .order_by(LogisticsPlanPreloadItem.created_at.asc())
        ).all()
    )


def _build_preload_read(db: Session, preload: LogisticsPlanPreload) -> PlanningPreloadRead:
    items = [
        PlanningPreloadItemRead.model_validate(item)
        for item in _get_preload_items(db, preload_id=preload.id)
    ]
    return PlanningPreloadRead(
        id=preload.id,
        tenant_id=preload.tenant_id,
        warehouse_id=preload.warehouse_id,
        branch_id=preload.branch_id,
        preload_date=preload.preload_date,
        status=preload.status,
        notes=preload.notes,
        created_by=preload.created_by,
        created_at=preload.created_at,
        updated_at=preload.updated_at,
        items=items,
    )


def build_preload_read(db: Session, preload: LogisticsPlanPreload) -> PlanningPreloadRead:
    return _build_preload_read(db, preload)


def list_preloads(
    db: Session,
    *,
    tenant_id: str,
    allowed_warehouse_ids: tuple[str, ...] | None,
    warehouse_id: str | None = None,
    preload_date: date | None = None,
    status: str | None = None,
) -> list[PlanningPreloadRead]:
    stmt = select(LogisticsPlanPreload).where(LogisticsPlanPreload.tenant_id == tenant_id)
    if warehouse_id is not None:
        stmt = stmt.where(LogisticsPlanPreload.warehouse_id == warehouse_id)
    if preload_date is not None:
        stmt = stmt.where(LogisticsPlanPreload.preload_date == preload_date)
    if status is not None:
        stmt = stmt.where(LogisticsPlanPreload.status == status)
    if allowed_warehouse_ids is not None:
        stmt = stmt.where(LogisticsPlanPreload.warehouse_id.in_(allowed_warehouse_ids))
    preloads = list(
        db.scalars(
            stmt.order_by(
                LogisticsPlanPreload.preload_date.desc(), LogisticsPlanPreload.created_at.desc()
            )
        ).all()
    )
    return [_build_preload_read(db, preload) for preload in preloads]


def get_preload(db: Session, *, tenant_id: str, preload_id: str) -> LogisticsPlanPreload | None:
    return db.scalar(
        select(LogisticsPlanPreload).where(
            LogisticsPlanPreload.id == preload_id,
            LogisticsPlanPreload.tenant_id == tenant_id,
        )
    )


def plan_order(
    db: Session,
    *,
    order: LogisticsOrder,
    payload: PlanningPlanOrderRequest,
    action_context: LogisticsActionContext,
) -> PlanningPlanOrderResult:
    if order.warehouse_id is None:
        raise ValueError("Order must have a warehouse to be planned")
    items = list(
        db.scalars(
            select(LogisticsOrderItem)
            .where(LogisticsOrderItem.order_id == order.id)
            .order_by(LogisticsOrderItem.created_at.asc())
        ).all()
    )
    product_ids = {item.product_id for item in items if item.product_id is not None}
    if len(product_ids) != len(items):
        raise ValueError("All order items must have product_id before planning")

    actual = _stock_actual_by_product(
        db,
        tenant_id=order.tenant_id,
        warehouse_id=order.warehouse_id,
        product_ids=product_ids,
    )
    committed = _stock_comprometido_by_product(
        db,
        tenant_id=order.tenant_id,
        warehouse_id=order.warehouse_id,
        product_ids=product_ids,
        exclude_order_id=order.id,
    )
    planned = _stock_planificado_by_product(
        db,
        tenant_id=order.tenant_id,
        warehouse_id=order.warehouse_id,
        product_ids=product_ids,
    )
    available_map = {
        product_id: actual.get(product_id, 0)
        - committed.get(product_id, 0)
        - planned.get(product_id, 0)
        for product_id in product_ids
    }

    if payload.mode in {"all", "full"} and not payload.permit_without_stock:
        for item in items:
            assert item.product_id is not None
            pending = max(float(item.quantity_requested) - float(item.quantity_planned), 0)
            if available_map.get(item.product_id, 0) < pending:
                raise ValueError("Insufficient stock for full planning mode")

    updated_items: list[OrderItemRead] = []
    for item in items:
        assert item.product_id is not None
        pending = max(float(item.quantity_requested) - float(item.quantity_planned), 0)
        if pending <= 0:
            updated_items.append(OrderItemRead.model_validate(item))
            continue
        available = available_map.get(item.product_id, 0)
        if payload.mode in {"all", "full"}:
            planned_now = pending if (available >= pending or payload.permit_without_stock) else 0
        else:
            planned_now = (
                pending if payload.permit_without_stock else min(pending, max(available, 0))
            )
        item.quantity_planned = float(item.quantity_planned) + planned_now
        if item.quantity_planned > 0:
            item.status = 1
        db.add(item)
        available_map[item.product_id] = available - planned_now
        updated_items.append(OrderItemRead.model_validate(item))

    order.status = "PLANIFICADO"
    db.add(order)
    db.flush()
    audit_logistics_action(
        db,
        context=action_context,
        branch_id=order.branch_id,
        action="planning.plan_order",
        entity_type="order",
        entity_id=order.id,
        details={"mode": payload.mode, "warehouse_id": order.warehouse_id},
    )
    return PlanningPlanOrderResult(
        order_id=order.id, mode=payload.mode, updated_items=updated_items
    )


def generate_preload(
    db: Session,
    *,
    tenant_id: str,
    created_by: str,
    payload: PlanningGeneratePreloadRequest,
    action_context: LogisticsActionContext,
) -> PlanningPreloadRead:
    warehouse = db.scalar(
        select(LogisticsWarehouse).where(
            LogisticsWarehouse.id == payload.warehouse_id,
            LogisticsWarehouse.tenant_id == tenant_id,
        )
    )
    if warehouse is None:
        raise LookupError("Warehouse not found")
    existing = db.scalar(
        select(LogisticsPlanPreload).where(
            LogisticsPlanPreload.tenant_id == tenant_id,
            LogisticsPlanPreload.warehouse_id == payload.warehouse_id,
            LogisticsPlanPreload.preload_date == payload.preload_date,
            LogisticsPlanPreload.status.in_(ACTIVE_PRELOAD_STATUSES),
        )
    )
    if existing is not None:
        raise ValueError("An active preload already exists for the selected date and warehouse")

    order_stmt = select(LogisticsOrder).where(
        LogisticsOrder.tenant_id == tenant_id,
        LogisticsOrder.warehouse_id == payload.warehouse_id,
        LogisticsOrder.status.in_(OPEN_ORDER_STATUSES),
    )
    if payload.order_ids:
        order_stmt = order_stmt.where(LogisticsOrder.id.in_(payload.order_ids))
    orders = list(db.scalars(order_stmt).all())
    preload = LogisticsPlanPreload(
        tenant_id=tenant_id,
        warehouse_id=payload.warehouse_id,
        branch_id=warehouse.branch_id,
        preload_date=payload.preload_date,
        status="PENDIENTE",
        notes=payload.notes,
        created_by=created_by,
    )
    db.add(preload)
    db.flush()
    created_items = 0
    for order in orders:
        items = list(
            db.scalars(
                select(LogisticsOrderItem).where(LogisticsOrderItem.order_id == order.id)
            ).all()
        )
        for item in items:
            if item.product_id is None or float(item.quantity_planned) <= 0:
                continue
            db.add(
                LogisticsPlanPreloadItem(
                    tenant_id=tenant_id,
                    preload_id=preload.id,
                    order_item_id=item.id,
                    product_id=item.product_id,
                    product_name=item.product_name,
                    quantity_planned=float(item.quantity_planned),
                    quantity_loaded=0,
                )
            )
            created_items += 1
    if created_items == 0:
        raise ValueError("No planned order items available to generate preload")
    db.flush()
    audit_logistics_action(
        db,
        context=action_context,
        branch_id=warehouse.branch_id,
        action="planning.preload.generate",
        entity_type="preload",
        entity_id=preload.id,
        details={
            "warehouse_id": preload.warehouse_id,
            "preload_date": preload.preload_date.isoformat(),
        },
    )
    emit_logistics_event(
        db,
        context=action_context,
        branch_id=warehouse.branch_id,
        event_name="logistics.planning.preload_generated",
        entity_type="preload",
        entity_id=preload.id,
        payload={
            "warehouse_id": preload.warehouse_id,
            "preload_date": preload.preload_date.isoformat(),
        },
    )
    return _build_preload_read(db, preload)


def accept_preload(
    db: Session,
    *,
    preload: LogisticsPlanPreload,
    action_context: LogisticsActionContext,
) -> PlanningPreloadActionResult:
    if preload.status != "PENDIENTE":
        raise ValueError("Only pending preloads can be accepted")
    movement = LogisticsMovement(
        tenant_id=preload.tenant_id,
        branch_id=preload.branch_id,
        movement_type="TR",
        warehouse_id=preload.warehouse_id,
        status="PENDIENTE",
        notes=preload.notes,
        created_by=action_context.actor_user_id,
    )
    db.add(movement)
    db.flush()
    for preload_item in _get_preload_items(db, preload_id=preload.id):
        db.add(
            LogisticsMovementItem(
                movement_id=movement.id,
                cylinder_id=None,
                product_id=preload_item.product_id,
                product_name=preload_item.product_name,
                quantity_in=0,
                quantity_out=float(preload_item.quantity_planned),
                quantity=0,
                quantity_planned=float(preload_item.quantity_planned),
                item_status="P",
                notes="Generated from preload acceptance",
            )
        )
    preload.status = "ACEPTADA"
    db.add(preload)
    db.flush()
    audit_logistics_action(
        db,
        context=action_context,
        branch_id=preload.branch_id,
        action="planning.preload.accept",
        entity_type="preload",
        entity_id=preload.id,
        details={"movement_id": movement.id, "warehouse_id": preload.warehouse_id},
    )
    emit_logistics_event(
        db,
        context=action_context,
        branch_id=preload.branch_id,
        event_name="logistics.planning.preload_accepted",
        entity_type="preload",
        entity_id=preload.id,
        payload={"movement_id": movement.id, "warehouse_id": preload.warehouse_id},
    )
    return PlanningPreloadActionResult(
        preload=_build_preload_read(db, preload),
        movement=MovementRead.model_validate(movement),
    )


def cancel_preload(
    db: Session,
    *,
    preload: LogisticsPlanPreload,
    action_context: LogisticsActionContext,
) -> PlanningPreloadRead:
    if preload.status == "CANCELADA":
        return _build_preload_read(db, preload)
    preload.status = "CANCELADA"
    db.add(preload)
    db.flush()
    audit_logistics_action(
        db,
        context=action_context,
        branch_id=preload.branch_id,
        action="planning.preload.cancel",
        entity_type="preload",
        entity_id=preload.id,
        details={"warehouse_id": preload.warehouse_id},
    )
    return _build_preload_read(db, preload)

File: plugins/logistics/frontend/api/planning.ts
// Auto-generado por split_api.py
import { API_PREFIX, withQuery } from "./_shared";
import { LogisticsOrderItem } from "./orders";
import { StockBalancePageRead } from "./_shared";
import { LogisticsMovement } from "./movements";
import { apiRequest } from "@systutor/shell/api/client";

export type PlanningStockSummaryItem = {
  product_id: string;
  product_name: string;
  warehouse_id: string;
  stock_actual: number;
  stock_comprometido: number;
  stock_planificado: number;
  stock_disponible: number;
  coverage_status: string;
};

export type PlanningStockOwnerBalanceItem = {
  product_id: string;
  product_name: string;
  warehouse_id: string;
  quantity: number;
};

export type PlanningPendingOrderItem = {
  order_item_id: string;
  product_id: string | null;
  product_name: string;
  quantity_requested: number;
  quantity_planned: number;
  quantity_pending: number;
  stock_disponible: number;
  coverage_status: string;
};

export type PlanningPendingOrder = {
  order_id: string;
  customer_id: string | null;
  customer_name: string;
  warehouse_id: string | null;
  status: string;
  coverage_status: string;
  items: PlanningPendingOrderItem[];
};

export type PlanningPlanOrderResult = {
  order_id: string;
  mode: string;
  updated_items: LogisticsOrderItem[];
};

export type PlanningPreloadItem = {
  id: string;
  tenant_id: string;
  preload_id: string;
  order_item_id: string;
  product_id: string;
  product_name: string | null;
  quantity_planned: number;
  quantity_loaded: number;
  created_at: string;
};

export type PlanningPreload = {
  id: string;
  tenant_id: string;
  warehouse_id: string;
  branch_id: string | null;
  preload_date: string;
  status: string;
  notes: string | null;
  created_by: string;
  created_at: string;
  updated_at: string;
  items: PlanningPreloadItem[];
};

export type PlanningPreloadActionResult = {
  preload: PlanningPreload;
  movement: LogisticsMovement | null;
};

export type PlanningExpectedLoadSummary = {
  items: {
    product_id: string;
    product_name: string;
    sku: string | null;
    quantity: number;
    adr_required: boolean;
    unit_weight_kg: number | null;
    total_weight_kg: number | null;
  }[];
  total_products: number;
  total_units: number;
  total_weight_kg: number | null;
};

export type PlanningReservation = {
  id: string;
  tenant_id: string;
  branch_id: string | null;
  vehicle_id: string;
  vehicle_plate: string;
  origin_warehouse_id: string;
  origin_warehouse_name: string;
  planned_start_at: string;
  planned_end_at: string;
  expected_load_summary: PlanningExpectedLoadSummary;
  expected_weight_total: number | null;
  expected_volume_total: number | null;
  service_type: string | null;
  route_id: string | null;
  driver_id: string | null;
  driver_name: string | null;
  customer_ids: string[];
  address_ids: string[];
  adr_required: boolean;
  notes: string | null;
  status: string;
  conflict_reason: string | null;
  permit_override: boolean;
  override_reason: string | null;
  linked_session_id: string | null;
  actual_start_at: string | null;
  actual_end_at: string | null;
  actual_load_summary: PlanningExpectedLoadSummary | null;
  created_at: string;
  updated_at: string;
};

export type PlanningReservationPayload = {
  vehicle_id: string;
  origin_warehouse_id: string;
  planned_start_at: string;
  planned_end_at: string;
  expected_load_summary: PlanningExpectedLoadSummary;
  expected_weight_total?: number | null;
  expected_volume_total?: number | null;
  service_type?: string | null;
  route_id?: string | null;
  driver_id?: string | null;
  customer_ids?: string[];
  address_ids?: string[];
  adr_required?: boolean;
  notes?: string | null;
  permit_override?: boolean;
  override_reason?: string | null;
};

export function getPlanningStock(warehouse_id?: string) {
  return apiRequest<PlanningStockSummaryItem[]>(
    withQuery(`${API_PREFIX}/planning/stock`, { warehouse_id })
  );
}

export function postPlanningStockSummary(payload: { warehouse_id: string; product_ids?: string[] }) {
  return apiRequest<PlanningStockSummaryItem[]>(`${API_PREFIX}/planning/stock/summary`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function listPlanningStockBalances(
  warehouse_id: string,
): Promise<PlanningStockOwnerBalanceItem[]> {
  const limit = 200;
  const items: PlanningStockOwnerBalanceItem[] = [];
  let offset = 0;
  let total = 0;

  do {
    const page = await apiRequest<StockBalancePageRead>(
      withQuery("/api/v1/plugins/stock/balance", {
        warehouse_id,
        limit,
        offset,
      }),
    );
    items.push(
      ...page.items.map((item) => ({
        product_id: item.product_id,
        product_name: item.product_name,
        warehouse_id: item.warehouse_id,
        quantity: item.quantity,
      })),
    );
    total = page.total;
    offset += page.limit;
  } while (items.length < total);

  return items;
}

export function listPlanningPendingOrders(warehouse_id?: string) {
  return apiRequest<PlanningPendingOrder[]>(
    withQuery(`${API_PREFIX}/planning/pending-orders`, { warehouse_id }),
  );
}

export function postPlanOrder(orderId: string, payload: { mode: string; permit_without_stock?: boolean }) {
  return apiRequest<PlanningPlanOrderResult>(`${API_PREFIX}/planning/plan-order/${orderId}`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function generatePreload(payload: { warehouse_id: string; preload_date: string; order_ids?: string[]; notes?: string }) {
  return apiRequest<PlanningPreload>(`${API_PREFIX}/planning/generate-preload`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function listPreloads(warehouse_id?: string) {
  return apiRequest<PlanningPreload[]>(withQuery(`${API_PREFIX}/planning/preloads`, { warehouse_id }));
}

export function getPreload(preloadId: string) {
  return apiRequest<PlanningPreload>(`${API_PREFIX}/planning/preloads/${preloadId}`);
}

export function acceptPreload(preloadId: string) {
  return apiRequest<PlanningPreloadActionResult>(`${API_PREFIX}/planning/preloads/${preloadId}/accept`, {
    method: "POST",
  });
}

export function cancelPreload(preloadId: string) {
  return apiRequest<PlanningPreload>(`${API_PREFIX}/planning/preloads/${preloadId}/cancel`, {
    method: "POST",
  });
}

export function listPlanningReservations(filters: {
  start?: string;
  end?: string;
  vehicle_id?: string;
} = {}) {
  return apiRequest<PlanningReservation[]>(withQuery(`${API_PREFIX}/planning/reservations`, filters));
}

export function createPlanningReservation(payload: PlanningReservationPayload) {
  return apiRequest<PlanningReservation>(`${API_PREFIX}/planning/reservations`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function updatePlanningReservation(reservationId: string, payload: Partial<PlanningReservationPayload>) {
  return apiRequest<PlanningReservation>(`${API_PREFIX}/planning/reservations/${reservationId}`, {
    method: "PATCH",
    body: JSON.stringify(payload),
  });
}

export function activatePlanningReservation(reservationId: string) {
  return apiRequest<PlanningReservation>(`${API_PREFIX}/planning/reservations/${reservationId}/activate`, {
    method: "POST",
  });
}

export function cancelPlanningReservation(reservationId: string) {
  return apiRequest<PlanningReservation>(`${API_PREFIX}/planning/reservations/${reservationId}/cancel`, {
    method: "POST",
  });
}
